### DISCUSSION TITLE
Help with LoRaWAN implementation on Raspberry Pi 5 with the Waveshare SX1262 HAT

### DISCUSSION THREAD

**Discussion by Miliks** [jgromes/RadioLib#1187] [Q&A]
i was trying to understand how i can implement the NonArduino code on my Raspberry Pi device. I cloned the repo on RPi and executed $ cmake ..
$ sudo make install in the example folder without any errors. But when i try to execute gcc -o RaspberryPi main.cpp it start to give me a lot of errors for the header files. i copied all the folders from include RadioLib file into the NonArduino folder, but now i'm getting an errors with the "undefined references SX1262, PhysicalLayer etc"
![shared image (4)](https://github.com/user-attachments/assets/b6035e63-319c-4a6e-8398-35748ae21ff1)
could you please help me!

**Comment by jgromes** [comment]
> executed $ cmake .. $ sudo make install in the example folder

That doesn't sound right, why in the examples folder? You can take a look at the CI, which installs the library on a Raspberry Pi runner and then performs some tests on it, for example here: https://github.com/jgromes/RadioLib/actions/runs/10334213403/job/28607425921

And of course there's the wiki ;) https://github.com/jgromes/RadioLib/wiki/Porting-to-non-Arduino-Platforms

**Reply by Miliks** [reply]
i made it again with the root directory were all the RadioLib files are located after that i run also a ,/buold.sh without any errors. So now i suppose to run gcc -o Filename main.cpp? but running it i'm getting the same errors reported in the first request... what i'm doing wrong, please help me?

**Reply by jgromes** [reply]
I don't know what exactly you're trying to do here, but the usualy workflow is following:

1. **Install the library**. You did that by `sudo make install` after CMake has been configured, just like any other library that uses CMake. You now have RadioLib.a installed in `/usr/local/lib/aarch64-linux-gnu/` and headers in `/usr/local/include/RadioLib`.
2. **Build your source code**. Use whatever build system you like - Makefiles, CMake, meson ... or call GCC manually if you like, that is completely up to you. In the examples, I have opted to use CMake (because I'm too lazy to write down all the calls to compiler, linker etc., that is why CMake exists). This is why the examples also have a `CMakeLists.txt` file. The `build.sh` script calls `cmake` and then `make`. Its output is an executable program `build/rpi-sx1261`. No need to call gcc manually.

I would suggest to take a look at how CMake is used in general.

**Reply by Miliks** [reply]
thank you for the patience, i'm not very familiar with the C++ development. I need to integrate the SX1262 Raspberry HAT from Waveshare to make a Raspberry work as LoRaWAN node. The Waveshare do not provide any support, keep pointing me to the Wiki, where only LoRa protocol is provided. So the only siutable library i found in the net is yours. 
So i'd like to find the way to use it to run a code to send a messages out from Raspberry using LoRaWAN protocol.
Step 1 - install library seems to do not go as it expected, because i do not have RadioLib.o in the path you specify, moreover the aarch-linux-gnu folder is located on /urs/lib/ and not /usr/local/lib. Does it matter?. the headers are where they suppose to be.
Running the build looks like created the executable file rpi-sx1261. So i run it and get an error [SX1261] Initializing ... failed, code -2. i will try to figure out what is it mean. But probably it could be caused by the problems i listed above? 
i have another question, do u know if a pin layout for the SX1261 e 1262 are the same? since they referenced together i assume so2, but just to be sure...

**Reply by Miliks** [reply]
Just to let you know what i've tried to change the pin mapping in the code following the pinout indicated on Waveshare website
![immagine (6)](https://github.com/user-attachments/assets/59616263-6c47-43ba-b5f8-fe77644b3efd)
SX1262 radio = new Module(hal, 21, 16, 18, RADIOLIB_NC);
meanwhile i also found in the Sparkfun datasheet (https://cdn.sparkfun.com/assets/6/b/5/1/4/SX1262_datasheet.pdf) for the chip SX1262 the following pinout: 
![image (3)](https://github.com/user-attachments/assets/9f20bdf4-738f-4337-a4f1-1917daffddef)
so i tryid to change the code like this:
SX1262 radio = new Module(hal, 19, 13, 15, RADIOLIB_NC);
but still getting Initializing ... failed, code -2. 
can you give me any suggestion?

**Reply by jgromes** [reply]
> i do not have RadioLib.o in the path you specify

Why do you expect a .o output? If you install the library as statically linked, it will be `libRadioLib.a`, if you install it as dynamically linked, it will be `libRadioLib.so`.

> Running the build looks like created the executable file rpi-sx1261

Exactly as the CMakeLists file specifies.

>But probably it could be caused by the problems i listed above?

No, the program runs now, so the library is installed correctly and the exectuable is correctly compiled and linked.

> SX1262 radio = new Module(hal, 21, 16, 18, RADIOLIB_NC);
> so i tryid to change the code like this:
> SX1262 radio = new Module(hal, 19, 13, 15, RADIOLIB_NC);
> but still getting Initializing ... failed, code -2.

First, you have to specify the BUSY pin as well, it will not work without it. Next, you are mixing together pin numbers of the SX126x package and the Raspberry Pi GPIO numbers, which the program expects. I happen to be using the same HAT, so I know for a fact that the correct pin assigment is `Module(hal, 21, 16, 18, 20);`. I'm guessing you have something wrong in your HAL, but since you haven't posted that, I can't be sure.

Look - you seem to missing a lot of basics in terms of writing, building and running C/C++ programs on Linux. You need to read up on libraries, the build process, configuration (via CMake for example) etc. before you jump into the more advanced stuff. This library can do a lot of the heavy lifting for you, you don't have to write the SX1262 and LoRaWAN driver from scratch. But the application that you want and its functionality will not write itself (nor will strangers on the Internet do it, for that matter). ;)

**Reply by Miliks** [reply]
Thanks a lot, i will try to get better knowledge on C/C++ programming meanwhile and try the changes you suggested. i didn't apply any changes on PiHal.h, if this is what you mean with something wrong with my HAL.

**Reply by HeadBoffin** [reply]
> Next, you are mixing together pin numbers of the SX126x package and the Raspberry Pi GPIO numbers, which the program expects.

For the sake of completeness as I've seen this misconception only a few times before, here's a longer explanation. The pins on the actual physical chip, the SX1262, are connected to pins on the RPi header which are the ones you need to use - there is no method or concept for a pin on the Pi to know which pin on a chip on a HAT it is connected to. So it would be worth you looking at the [schematic](https://files.waveshare.com/wiki/SX1262-XXXM-LoRaWAN-GNSS-HAT/SX1262_XXXM_LoRaWAN_HAT_sch.pdf) so you can see where these pin numbers come from as getting the pins right is something we all struggle with all the time when using a board for the first dozen or so times, **all of us**!

On the schematic you will see that the SX1262 NSS (negative slxxx select) on it's pin 19 has a label on it of CS (chip select) which links it to the header layout (top right) where CS is attached to the physical pin 40. This pin 40 is connected to the main Pi MCU which is from Broadcom, aka BCM, so the code that runs on the Pi needs to reference the pin number for the Pi's MCU which is 21. 

There are two additional things to know. 

1. The older SX127X range had a DIO0, on the SX126X range this is now called BUSY. This pin and the DIO1 indicate the status of the radio without having to read any information from its internal registers, all of which helps with the timing issues.
2. In the early days of the Pi, accessing the pins wasn't so easy and a library was created to help with it called WiringPi - which is why you sometimes see, like on the schematic, a reference to other pin numbers - hence the wPi.


Foot notes:

- HAT connections: if there was an ident EEPROM it could know the pin numbers to use, but there isn't one on the board so technically it can't officially be called a HAT, but that's a total other story.
- NSS (negative slxxx select): slxxx is a redaction - I'm not 100% politically correct but I see no reason not to evolve language, as it has done for hundreds of years, with the times as the prior references are problematic. More info [here](https://www.oshwa.org/a-resolution-to-redefine-spi-signal-names/).

**Comment by HeadBoffin** [comment]
I see that you have found your way over from the TTN forum. I'd be happy to pickup from there except we still don't know why you want to use a Pi 4/5 and for TTN use there is some hesitation amongst the community with Pi based devices as they can quickly become non-compliant due to timing issues as outlined. Also, that screen shot is a nightmare to read & can't be copied & pasted so please please post as text. It was also not made clear over on TTN which board you are using, so it would be useful to answer that as well please.

**Comment by HeadBoffin** [comment]
Given your self-professed lack of C knowledge, there are Python implementations around you could look at - from people on TTN who host them on GitHub.

**Comment by Miliks** [comment]
i don't mind at all to use the python code, but i didn't find anything useful for the HW combination i have (i would appreciate if you point me to some resources for python code for this boards combination). i use the raspberry pi to collect some data out from the sensors and do some calculation and was looking to be able to send the data from time to time from Raspberry Pi directly to LoRaWAN network. I bought the Raspberry Pi SX1262 LoRaWAN HAT from Waveshare as it was written to be LoRaWAN complaint.  From the TTN forum discussions i found this library as an only valid choice to implement what i want so far.

**Reply by HeadBoffin** [reply]
> I bought the Raspberry Pi SX1262 LoRaWAN HAT from Waveshare

Which one - please provide a link so we can be 100% clear you have the board that will work.

**Reply by Miliks** [reply]
https://www.waveshare.com/sx1262-lorawan-hat.htm - if you are saying what it is not comlaint then i'm sorry what Waveshare is saying the following on there website:  Combined with a LoRa gateway, it can be connected to servers such as TTN to build a LoRaWAN network. In addition to the basic LoRaWAN version, it also provides optional GNSS positioning function with GPS/BD support. So i was convinced what i'll fine using it.

**Reply by HeadBoffin** [reply]
> if you are saying what it is not comlaint then i'm sorry what Waveshare is saying the following on there website

I did not say it was not compliant. What was pointed out to you on TTN was that there were two boards from WaveShare and you hadn't confirmed which one you had. There is no point volunteers answering if you don't have the right hardware and it's not a big ask for you to confirm that you do. Given we answer questions for free, even for commercial businesses like Concept Reply, on a regular basis, allowing the well refined support process to run is quicker overall.

GitHub is littered with Python LoRaWAN implementations but I'd cross reference it against the ones discussed on the TTN forum so that you know its background.

Alternatively, review the materials linked to above for implementing RadioLib and report results. There are some prior issues & discussions on using the Pi that may provide pointers to assist.

**Reply by Miliks** [reply]
Sorry for not replying on TTN forum, i haven't notice the thread over there. At the moment i did it was already closed.

**Reply by HeadBoffin** [reply]
Searching for "python LoRaWAN" on both GitHub and TTN will cross match and come up with this as your best starting point:

https://github.com/BNNorman/dragino-1

Whilst I've not used it as I wouldn't use a Pi as a LW device, I do know the author and his output is solid, so this is as close to a recommendation as I can think of, other than the recommendation not to use a Pi and move your data processing code to a different MCU.

However **you** will need to convert the code to use the SX1262 with one of the many Python implementations available for it or get a different HAT.

**Reply by Miliks** [reply]
Thanks a lot for the suggestion and the deep inside on the pinout confusion in the previous post, this is really precious! I actually used already RAK Wisblock base with the RAK4631 module to receive messages from RPi5 and send them out through LoRaWAN network. But i thought it is overwhelming to use 2 processing unit to send the data, so when i found this module for RPi with the same radio i thought it would be more elegant solution to use it directly from RPi. I'll give it a try and get back here with the results.

**Reply by HeadBoffin** [reply]
I would really stick to the RAK4631 for a Pi - I've used the original RAK811 on Pi and if I need to use a Pi for whatever reason at present, I wire in a Seeed LoRa-E5 - all provide AT command LW. The power a Pi uses having a second MCU that can usually sleep at the sub 10uA range isn't usually that much of an imposition.

**Reply by Miliks** [reply]
Probably i'll follow your suggestion and stick with RAK.
At this point i do not understand how Waveshare can sale the HAT for RPi without it being a HAT.

i'll investigate deeper, but from the first try i did, when i tried to call Module(hal, 21,16, 18, 20) it gives me an error: "Error writing reading from pin 20: GPIO not allocated."
if i add this piece of code:
`int gpio_handle = lgGpiochipOpen(0); // Open GPIO chip 0 (usually the default)
if (gpio_handle < 0) {     std::cerr << "Failed to open GPIO chip: " << strerror(errno) << std::endl;  
   return -1; }
if (lgGpioClaimInput(gpio_handle, 0, 20) < 0) {    
 std::cerr << "Failed to set GPIO pin 20 as input: " << strerror(errno) << std::endl;   
  return -1; }`
it return me "Failed to set GPIO pin 20 as input: Device or resource busy"
Could it be because i'm using the RPi5 instead of RPi4?

**Reply by HeadBoffin** [reply]
> At this point i do not understand how Waveshare can sale the HAT for RPi without it being a HAT.

The HAT specification published by RPi requires a small EEPROM to identify what is connected which can then load any dtoverlay's that are required at boot - or just confirm configuration details. But many manufacturer's don't include the EEPROM as the HAT spec can't be enforced.

I see it as a sign of how diligent a manufacturer is - if I'm going to create a device for a project I'll happily prototype internally with any number of vendors, WaveShare, Heltec, LilyGo as well as the official development boards. But for production I'd be cautious about using modules that can't be proven to meet good practise.

The Pi5 introduces a standby power state so there may be some material changes to connectivity that mean the board doesn't work well with a Pi5. That said, anyone using a Pi5 for local data processing must need be doing something pretty serious which makes using a maker level radio somewhat incongruous and as I keep saying, has potential for timing errors.

So overall, I'd just stop with the cheap WaveShare on top of a nice powerful Pi5 and stick with the RAK.

For me, a Pi5 for IoT isn't my thing as I don't need that sort of horsepower at the edge so I'm definitely not going to risk my one Pi5 experimenting with the GPIO!

**Reply by jgromes** [reply]
Probably worth pointing out that I do not own a RPi 5 and all of my development in this regard has been done on an old 3B+:

```
pi@raspberrypi:~ $ cat /proc/cpuinfo
(...)
Revision        : a02082
Serial          : 00000000a8912e2b
Model           : Raspberry Pi 3 Model B Rev 1.2
```

So it is possible that the lgpio library (https://abyz.me.uk/lg/lgpio.html) which is used in the example HAL to control the GPIO/SPI is not compatible with RPi 5.

**Reply by Miliks** [reply]
have u ever try it on RPi4B+?

**Reply by Miliks** [reply]
> > At this point i do not understand how Waveshare can sale the HAT for RPi without it being a HAT.
> 
> The HAT specification published by RPi requires a small EEPROM to identify what is connected which can then load any dtoverlay's that are required at boot - or just confirm configuration details. But many manufacturer's don't include the EEPROM as the HAT spec can't be enforced.
> 
> I see it as a sign of how diligent a manufacturer is - if I'm going to create a device for a project I'll happily prototype internally with any number of vendors, WaveShare, Heltec, LilyGo as well as the official development boards. But for production I'd be cautious about using modules that can't be proven to meet good practise.
> 
> The Pi5 introduces a standby power state so there may be some material changes to connectivity that mean the board doesn't work well with a Pi5. That said, anyone using a Pi5 for local data processing must need be doing something pretty serious which makes using a maker level radio somewhat incongruous and as I keep saying, has potential for timing errors.
> 
> So overall, I'd just stop with the cheap WaveShare on top of a nice powerful Pi5 and stick with the RAK.
> 
> For me, a Pi5 for IoT isn't my thing as I don't need that sort of horsepower at the edge so I'm definitely not going to risk my one Pi5 experimenting with the GPIO!

ok, thanks for the clarification. Yes, it was just a prototype for the proof of concept and it fails. I'm not stick with the RPi5 or 4, i was just hoping to use then since there performance were greater, I was trying RPi Pico with the Waveshare LoRaWAN HAT and have no problem on developing on it but hey have limited amount of pins to connect sensors., could't imaging all this difficulties changing a board.

**Reply by HeadBoffin** [reply]
> have u ever try it on RPi4B+?

> all of my development in this regard has been done on an old 3B+:

Looking unlikely.

To spin back round to the very start, again, what is it that the Pi does that can't be done on another MCU?

Otherwise we are well in to flogging a dead horse territory - if it has to be a Pi and with the options available to you in terms of resources & knowledge, just go RAK! If it's going to be produced in volume, then commission someone to sort it out.

**Reply by Miliks** [reply]
Good morning everyone, I would like to give some updates on what has been done:
Given the problem with the GPIO pins of the Raspberry Pi 5, I tested the GPIO pins used by the SX1262 with various libraries and found that pin 20 was not available, even when accessed through a simple script to identify it as input/output.
I managed to enable all the GPIO pins used for SPI by modifying the /boot/firmware/config.txt file
After retesting the pins, all of them could be accessed.
I then reconnected the SX1262 to the Raspberry Pi 5, and now the radio module is finally initialized correctly.
Based on the examples provided in example/Lorawan, I generated a new main.cpp and config.h file to establish a LoRaWAN connection.
The main.cpp works fine up until the connection to the LoRaWAN network (in node.activateOTAA()).
In LoraWan.cpp, it stops at this->phyLayer->transmit(joinRequestMsg, RADIOLIB_LORAWAN_JOIN_REQUEST_LEN); with state -5 . From RadioLib seems to be (
#define | RADIOLIB_ERR_TX_TIMEOUT   (-5)
-- | Timed out waiting for transmission finish.)
i can see from the TTN server what the node send a joint request and it's accepted. I didn't change anything in the files for the network configuration part except the registration keys and the frequency to be compliant with our region. Is there anything what you can suggest to change to fix this error? 
[main.txt](https://github.com/user-attachments/files/17060017/main.txt)
[config.txt](https://github.com/user-attachments/files/17060018/config.txt)

**Reply by jgromes** [reply]
Tx completion is signaled via the IRQ pin. If the signal doesn't arrive within 500% of the expected time-on-air of the packet, RadioLib will return the error you have reported. So two things can be going wrong:

1. There is no connection to the IRQ.
2. The time measurement is not working as expected. Thsi function is provided by the HAL (specifically by `millis()`).

**Comment by philhuen** [comment]
RPi4B+ and 2W both  have the -> Initializing ... failed, code -707

Anyone able to use this HAT successfully? Did try SX1262 radio = new Module(hal,21, 16, 18, 20);

mi@localhost:/RadioLib-master/examples/NonArduino/Raspberry $ sudo bash build.sh
-- Configuring done
-- Generating done
-- Build files have been written to: /RadioLib-master/examples/NonArduino/Raspberry/build
[ 50%] Building CXX object CMakeFiles/rpi-sx1261.dir/main.cpp.o
[100%] Linking CXX executable rpi-sx1261
[100%] Built target rpi-sx1261
   text    data     bss     dec     hex filename
  60579     748     420   61747    f133 build/rpi-sx1261
mi@localhost:/RadioLib-master/examples/NonArduino/Raspberry $ sudo nano main.cpp
mi@localhost:/RadioLib-master/examples/NonArduino/Raspberry $ ./build/rpi-sx1261
[SX1261] Initializing ... failed, code -707


mi@localhost:/RadioLib-master/examples/NonArduino/Raspberry $ cat /sys/kernel/debug/gpio
gpiochip0: GPIOs 512-569, parent: platform/fe200000.gpio, pinctrl-bcm2711:
 gpio-512 (ID_SDA              )
 gpio-513 (ID_SCL              )
 gpio-514 (GPIO2               )
 gpio-515 (GPIO3               )
 gpio-516 (GPIO4               )
 gpio-517 (GPIO5               )
 gpio-518 (GPIO6               )
 gpio-519 (GPIO7               |spi0 CS1            ) out hi ACTIVE LOW
 gpio-520 (GPIO8               |spi0 CS0            ) out hi ACTIVE LOW
 gpio-521 (GPIO9               )
 gpio-522 (GPIO10              )
 gpio-523 (GPIO11              )
 gpio-524 (GPIO12              )
 gpio-525 (GPIO13              )
 gpio-526 (GPIO14              )
 gpio-527 (GPIO15              )
 gpio-528 (GPIO16              )
 gpio-529 (GPIO17              )
 gpio-530 (GPIO18              )
 gpio-531 (GPIO19              )
 gpio-532 (GPIO20              )
 gpio-533 (GPIO21              )
 gpio-534 (GPIO22              )
 gpio-535 (GPIO23              )
 gpio-536 (GPIO24              )
 gpio-537 (GPIO25              )
 gpio-538 (GPIO26              )
 gpio-539 (GPIO27              )
 gpio-540 (RGMII_MDIO          )
 gpio-541 (RGMIO_MDC           )
 gpio-542 (CTS0                )
 gpio-543 (RTS0                )
 gpio-544 (TXD0                )
 gpio-545 (RXD0                )
 gpio-546 (SD1_CLK             )
 gpio-547 (SD1_CMD             )
 gpio-548 (SD1_DATA0           )
 gpio-549 (SD1_DATA1           )
 gpio-550 (SD1_DATA2           )
 gpio-551 (SD1_DATA3           )
 gpio-552 (PWM0_MISO           )
 gpio-553 (PWM1_MOSI           )
 gpio-554 (STATUS_LED_G_CLK    |ACT                 ) out lo
 gpio-555 (SPIFLASH_CE_N       )
 gpio-556 (SDA0                )
 gpio-557 (SCL0                )
 gpio-558 (RGMII_RXCLK         )
 gpio-559 (RGMII_RXCTL         )
 gpio-560 (RGMII_RXD0          )
 gpio-561 (RGMII_RXD1          )
 gpio-562 (RGMII_RXD2          )
 gpio-563 (RGMII_RXD3          )
 gpio-564 (RGMII_TXCLK         )
 gpio-565 (RGMII_TXCTL         )
 gpio-566 (RGMII_TXD0          )
 gpio-567 (RGMII_TXD1          )
 gpio-568 (RGMII_TXD2          )
 gpio-569 (RGMII_TXD3          )

gpiochip1: GPIOs 570-577, parent: platform/soc:firmware:gpio, raspberrypi-exp-gpio, can sleep:
 gpio-570 (BT_ON               |shutdown            ) out hi
 gpio-571 (WL_ON               )
 gpio-572 (PWR_LED_OFF         |PWR                 ) out lo ACTIVE LOW
 gpio-573 (GLOBAL_RESET        )
 gpio-574 (VDD_SD_IO_SEL       |vdd-sd-io           ) out hi
 gpio-575 (CAM_GPIO            |cam1_regulator      ) out lo
 gpio-576 (SD_PWR_ON           |regulator-sd-vcc    ) out hi
 gpio-577 (SD_OC_N             )
mi@localhost:/RadioLib-master/examples



Here is the code:
mi@localhost:/RadioLib-master/examples/NonArduino/Raspberry $ cat  main.cpp
/*
   RadioLib Non-Arduino Raspberry Pi Example

   This example shows how to use RadioLib without Arduino.
   In this case, a Raspberry Pi with WaveShare SX1302 LoRaWAN Hat
   using the lgpio library
   https://abyz.me.uk/lg/lgpio.html

   Can be used as a starting point to port RadioLib to any platform!
   See this API reference page for details on the RadioLib hardware abstraction
   https://jgromes.github.io/RadioLib/class_hal.html

   For full API reference, see the GitHub Pages
   https://jgromes.github.io/RadioLib/
*/

// include the library
#include "src/RadioLib.h"

// include the hardware abstraction layer
#include "PiHal.h"

// create a new instance of the HAL class
// use SPI channel 1, because on Waveshare LoRaWAN Hat,
// the SX1261 CS is connected to CE1
PiHal* hal = new PiHal(1);

// now we can create the radio module
// pinout corresponds to the Waveshare LoRaWAN Hat
// NSS pin:   7
// DIO1 pin:  17
// NRST pin:  22
// BUSY pin:  not connected
//SX1261 radio = new Module(hal, 7, 17, 22, RADIOLIB_NC);
SX1262 radio = new Module(hal,21, 16, 18, 20);


// the entry point for the program
int main(int argc, char** argv) {
  // initialize just like with Arduino
  printf("[SX1261] Initializing ... ");
  int state = radio.begin();
  if (state != RADIOLIB_ERR_NONE) {
    printf("failed, code %d\n", state);
    return(1);
  }
  printf("success!\n");

  // loop forever
  int count = 0;
  for(;;) {
    // send a packet
    printf("[SX1261] Transmitting packet ... ");
    char str[64];
    sprintf(str, "Hello World! #%d", count++);
    state = radio.transmit(str);
    if(state == RADIOLIB_ERR_NONE) {
      // the packet was successfully transmitted
      printf("success!\n");

      // wait for a second before transmitting again
      hal->delay(1000);

    } else {
      printf("failed, code %d\n", state);

    }

  }

  return(0);
}

**Reply by philhuen** [reply]
Fixed by src/modules/SX126x/SX1262.h

 tcxoVoltage = 0

int16_t begin(float freq = 868.0, float bw = 125.0, uint8_t sf = 9, uint8_t cr = 7, uint8_t syncWord = RADIOLIB_SX126X_SYNC_WORD_PRIVATE, int8_t power = 10, uint16_t preambleLength Length = 8, float tcxoVoltage = 0, bool useRegulatorLDO = false);//tcxoVoltage = 1.6

**Reply by jgromes** [reply]
@philhuen ... or you could read the documentation for error -707 which says:

```c++
/*!
  \brief SX126x failed to execute SPI command.
  Often this means that the module is trying to use TCXO while
  XTAL is connected (or vice versa). Make sure your crystal setup
  (e.g. TCXO reference voltage) matches your hardware by setting
  "tcxoVoltage" to 0 when using XTAL module, or to appropriate value
  when using TCXO module.
*/
#define RADIOLIB_ERR_SPI_CMD_FAILED                            (-707)
```

We even have a troublehooting guide entry for this, which shows two different way to select the correct oscillator, none of which requires editing the library source code: https://github.com/jgromes/RadioLib/wiki/Troubleshooting-Guide#sx126xsx128x-incorrect-oscillator
